Chkstat: Fix error reporting

mgerstner opened the pull request.

mgerstner committed:
chkstat regtests: catch bad error reporting for non-existent files

mgerstner committed:
chkstat: EntryProcessor: don't report errors for non-existent files

The situation of files simply not existing has been overlooked when
adding more rigorous error reporting.

FilippoBonazziSUSE commented on src/entryproc.cpp:
```diff
@@ -32,11 +32,13 @@ bool EntryProcessor::process(const bool have_proc) {
 
         if (!safeOpen()) {
             return false;
+        } else if (!m_fd.valid()) {
```
Could this check not rather be done in `Chkstat::processEntries`? Like you call `needToCheck` on the entry path before even entering `EntryProcessor::process`.

mgerstner replied:
Theoretically / semantically, yes, but we're investing a lot of effort here to `safeOpen()` the path. That would be made a mockery of if we e.g. simply `stat()` if the path exists or not in advance.

mgerstner replied:
By now it's also he case that Chkstat only applies policy i.e. what the configuration files and command line parameters dictate. EntryProcessor deals with the reality found on the file system. So from this POV I believe it is the right place to check this also semantically.

FilippoBonazziSUSE commented on src/entryproc.cpp:
```diff
@@ -175,8 +177,9 @@ bool EntryProcessor::safeOpen() {
+                    return false;
                 }
-                return false;
+                return true;
```
I'm confused here a bit. Why silently return true? Is there a situation where we can get this far on a non-existing file?

mgerstner replied:
Yes, because `errno == ENOENT`, maybe I should put it in an `else` clause instead, or generally refactor the `if`.-

mgerstner replied:
The problem right now is that it silently returns `false`, which is why we don't see any additional error diagnostic in the OBS buildlog. Instead we need to allow the caller to differentiate between fatal and non fatal errors. I also thought about introducing a state `enum` that better reflects the outcome of `safeOpen()` instead of coupling it with the status of `m_fd`.

FilippoBonazziSUSE replied:
It could be the right moment to add coherent error handling to cover all cases. It depends how long it would take, and how long this submission can wait.

mgerstner replied:
I don't think any case is missing currently, it's just not straightforward to see. I will try something using an `enum`.

FilippoBonazziSUSE replied:
Ok yes the additional commit makes sense.

FilippoBonazziSUSE reviewed (commented):
I thought comments appeared immediately when I post them. maybe I need to finish the review and they appear in bulk. Anyway let's continue the discussion tomorrow

mgerstner commented:
> I thought comments appeared immediately when I post them. maybe I need to finish the review and they appear in bulk. Anyway let's continue the discussion tomorrow

I believe you can choose whether to post comments right away or assemble them to submit in one go when finishing the review. But maybe I'm mixing this up with GitLab right now.

mgerstner committed:
chkstat: EntryProcessor: make error handling in safeOpen() clearer

There are three outcomes that safeOpen() can have:

- file has been opened and processing should continue.
- a problematic error condition has been found that has been diagnosed
  and should be reported via a non-zero exit status.
- a non-error condition has been found that prevents further processing.
  No diagnostic is output, exit status should not be changed.

To make this clearer use an enum to model these three outcomes. No
longer inspect the state of `m_fd` and `m_file_status` but keep all this
logic within `safeOpen()`.
